python/eos/tasks.py

```diff
@@ -0,0 +1,196 @@
+# vim: set sw=4 sts=4 et tw=120 :
+
+# Copyright (c) 2020-2021 Danny van Dyk
+#
+# This file is part of the EOS project. EOS is free software;
+# you can redistribute it and/or modify it under the terms of the GNU General
+# Public License version 2, as published by the Free Software Foundation.
+#
+# EOS is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+# Place, Suite 330, Boston, MA  02111-1307  USA
+
+import eos
+import logging
+import numpy as _np
+import os
+import pypmc
+
+def sample_mcmc(analysis_file, posterior, chain, base_directory='./', pre_N=150, preruns=3, N=1000, stride=5):
+    """
+    Samples from a named posterior PDF using Markov Chain Monte Carlo (MCMC) methods.
+
+    The output file will be stored in EOS_BASE_DIRECTORY/POSTERIOR/mcmc-IDX.
+
+    :param analysis_file: The name of the analysis file that describes the named posterior.
+    :type analysis_file: str
+    :param posterior: The name of the posterior PDF from which to draw the samples.
+    :type posterior: str
+    :param chain: The index assigned to the Markov chain. This value is used to seed the RNG for a reproducable analysis.
+    :type chain: int >= 0
+    :param base_directory: The base directory for the storage of data files. Can also be set via the EOS_BASE_DIRECTORY environment variable.
+    :type base_directory: str, optional
+    :param pre_N: The number of samples to be used for an adaptation in each prerun steps. These samples will be discarded.
+    :type pre_N: int, optional
+    :param preruns: The number of prerun steps, which ared used to adapt the MCMC proposal to the posterior.
+    :type preruns: int, optional
+    :param N: The number of samples to be stored in the output file. Defaults to 1000.
+    :type N: int, optional
+    :param stride: The ratio of samples drawn over samples stored. For every S samples, S - 1 will be discarded. Defaults to 5.
+    :type stride: int, optional
+    """
+
+    output_path = os.path.join(base_directory, posterior, 'mcmc-{:04}'.format(int(chain)))
+    _set_log_file(output_path, 'log')
+    _analysis_file = eos.AnalysisFile(analysis_file)
+    analysis = _analysis_file.analysis(posterior)
+    rng = _np.random.mtrand.RandomState(int(chain) + 1701)
```

Comment:
Should we set the seed as an optional parameter?

Reply:
Not at this point. I want calls to sample-mcmc to be reproducable, hence the chain-index based seed.

Reply:
If the argument is optional with a default value, the reproducibility is ensured, no ?